Automattic/sensei
# Add TypeScript support

### Changes proposed in this Pull Request

* Add typescript support to Sensei;
* Convert two files from JS to TypeScript;

### Testing instructions

1. Verify if the command `npm run lint-js` still catches linting issues on `.ts` files;
2. Verify if modifications to `.ts` files still appear on the frontend;
3. Verify if the command `npm run lint-types` can catch type errors in `.ts` files;

I'd suggest also "playing" a little bit with the implementation. E.g. by creating a React component using TypeScript and verifying how the type-check works there.

### Implementation notes

This PR uses Babel to transpile the code during the Webpack compilation, which means that it will NOT type-check the code while building the assets.

Instead, this PR implements an additional command, called `npm run lint-types`, to run the TypeScript compiler on the whole project. This command DOES NOT generate or modify any files, it just types checks and shows an error if any error is detected, so it is safe to run even during development. A good idea might be to run the compiler on the command `npm run lint-js`, but for now I preferred to have a separate command instead.

On IDEs and some code editors (like `vscode`), it should still detect any errors, thanks to the presence of the `tsconfig.json` on the root of the project.

For more details on why to use babel on webpack compilation and have an additional command instead, see: https://www.typescriptlang.org/docs/handbook/babel-with-typescript.html

Note that the `eslint` config was also adapted to transpile and analyze `.ts` files, which means that it should still catch linting errors on `.ts` files.
